pkg/kubectl/cmd/get/get.go

```diff
@@ -564,9 +572,17 @@ func (o *GetOptions) watch(f cmdutil.Factory, cmd *cobra.Command, args []string)
 	}
 
 	first := true
-	intr := interrupt.New(nil, w.Stop)
+	ctx, cancel := watchtools.ContextWithOptionalTimeout(context.Background(), o.Timeout)
+	defer cancel()
+	intr := interrupt.New(nil, cancel)
 	intr.Run(func() error {
-		_, err := watch.Until(0, w, func(e watch.Event) (bool, error) {
+		// FIXME: Switch to an informer (watchtools.UntilWithInformer) + dynamic client
+		// Pure WATCH will fail after some time but definitely on API timeout.
```

Comment:
I'm not sure we want to use an informer, or that we want to remove watch-only as it is today.  We can certainly retry the watch from the last index until such a time as we can't.  The "endless watch" behavior of an informer is a different type of command (wait or observe).

Reply:
["So I would agree that in
certain circumstances (like tests, where the order of operations matters)
that a retry watcher that watches from the last RV received and errors if
it can't continue is a useful tool, BUT it should be really clear that most
people shouldn't use this.  That may imply we want to put it in its own
package, and clearly document when it is appropriate to use it."](https://github.com/kubernetes/kubernetes/pull/50102#issuecomment-321387829)

I believe that informer behavior should be the default option in CLI commands, unless you need to have the order of events guaranteed and fail. Informer also gives you the same guarantee for as long as the RetryWatcher would work.

Why `kubectl get` needs to fail if that guarantee is broken and not to try recovery?


> The "endless watch" behavior of an informer is a different type of command (wait or observe).

I don't think this applies in general (unless you meant it only for the case of `kubectl get`). Take the example of oc rollout status. If you implement it by RetryWatcher there will always be an option that it can fail, no way around that. Users re-implementing those in `oc observe` is not an option.

Reply:
or if the true intent of watch-only is --since=now? I guess we could filter events in the watch loop lower than an RV or something like that

Reply:
I am open to extracting those into and replacing it with a linked issue going with retrying Until for now.

Reply:
> we could filter events in the watch loop lower than an RV

RV is opaque to clients... don't do client-side comparison

Reply:
To reach consensus sooner, I've dropped it from this PR and it uses only retry

Reply:
Actually it now uses only watchtools.UntilWithoutRetry because the unit tests are based on the unfortunate fact that we couldn't deal with watches being closed. Added a todo to replace it with retry (.Until) instead of an informer as asked.